### DISCUSSION TITLE
How to create a specific HID input device (for Windows) with Arduino-Pico?

### DISCUSSION THREAD

**Discussion by FerGT50** [earlephilhower/arduino-pico#1931] [Q&A] [answered]
Hello,
I would like to create a RP2040-based HID gaming device with a bunch of buttons, switches and dials (pots).

I assume I need to somehow "describe" its layout, values range etc. to Windows.

I studied https://arduino-pico.readthedocs.io/en/latest/usb.html and successfully run the Mouse example, but I am looking for more specific examples/docs, in particular about how to "describe" the specific device in order for Windows to properly recognize its controls.

Any hint?
Thank you very much

**Comment by earlephilhower** [comment]
Look at what the `Joystick` class provides.  If that's not enough, you'll need to use https://github.com/adafruit/Adafruit_TinyUSB_Arduino and build your own USB HID descriptor and report format in binary, register it, and then send reports in that format (like the existing Joystick/Mouse/etc. do).

**Reply by FerGT50** [reply]
Thank you, will look into it and follow up!

**Comment by FerGT50** [comment]
Hello, following up.
I have studied most TinyUSB examples; unfortunately I need to create a composite device which expose a Gamepad and a Serial port (CDC).

This case is not covered in any of the examples; at most, I have found HID_composite, which creates a composite HID device which acts as a Keyboard + Mouse.
Here, the HID descriptor is built using TinyUSB templates, like say

```
uint8_t const desc_hid_report[] =
{
  TUD_HID_REPORT_DESC_KEYBOARD( HID_REPORT_ID(RID_KEYBOARD) ),
  TUD_HID_REPORT_DESC_MOUSE   ( HID_REPORT_ID(RID_MOUSE) )
};
```

I hoped I could do something like

```
uint8_t const desc_hid_report[] =
{
  TUD_HID_REPORT_DESC_GAMEPAD( HID_REPORT_ID(RID_GAMEPAD) ),
  TUD_HID_REPORT_DESC_CDC( HID_REPORT_ID(RID_CDC) )  /* My hypotesis */
};
```
but unfortunately there is no template for CDC.
CDC-related examples are totally different: the virtual COM port is created using

`Adafruit_USBD_CDC USBSer1;`

which does not work along with the aforementioned custom HID report. :(

Any hint?

Many thanks

**Reply by earlephilhower** [reply, marked as answer]
You're going to need to make a compound USB device.  One HID node with 1 joystick report, and one CDC node.  The CDC stuff is not part of the HID report structure/etc.

See how the combined USB descriptor is made in the [core](https://github.com/earlephilhower/arduino-pico/blob/master/cores/rp2040/RP2040USB.cpp).  You then service the CDC callbacks and the HID callbacks separately.

You might also get a better response posting this question in the Adafruit TinyUSB repo since I'm pretty sure you're not the first person to try this.  :)

**Reply by FerGT50** [reply]
Thank you
